[SE-0461] Replace `@execution(...)` with `@concurrent` and `nonisolated(nonsending)`

xedin opened the pull request:
- Add `@concurrent` declaration and type attributes to replace `@execution(concurrent)`
- Add new `nonsending` modifier to `nonisolated` declaration modifier
- Make it possible to use `nonisolated(nonsending)` in type and expression contexts (modeled via `CallerIsolatedTypeRepr`)
- Replace `@execution(caller)` with `nonisolated(nonsending)` modifier
- Remove all of the references to `@execution` attribute throughout the codebase
- Remove `ExecutionAttribute` experimental flag

<!--
If this pull request is targeting a release branch, please fill out the
following form:
https://github.com/swiftlang/.github/blob/main/PULL_REQUEST_TEMPLATE/release.md?plain=1

Otherwise, replace this comment with a description of your changes and
rationale. Provide links to external references/discussions if appropriate.
If this pull request resolves any GitHub issues, link them like so:

  Resolves <link to issue>, resolves <link to another issue>.

For more information about linking a pull request to an issue, see:
https://docs.github.com/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[Diagnostics] Adjust `@execution(...)` diagnostic to take `DeclAttribute` or `StringRef`

It has been decided to split the attribute into `@concurrent` and
`nonisolated(nonsending`. Adjusting diagnostics to accept the attribute
makes the transition easier.

xedin committed:
[AST/ASTGen] Introduce `@concurrent` attribute to replace `@execution(concurrent)` spelling

xedin committed:
[Parse] Remove warning about `@concurrent` being alternative to `@Sendable`

xedin committed:
[AST/Sema] Support for `@concurrent` attribute in type context

xedin committed:
[AST/Sema] Replace `@execution(concurrent)` with `@concurrent`

xedin committed:
[AST/Sema] Intoduce `nonisolated(nonsending)` as a replacement for `@execution(caller)`

xedin committed:
[AST/Parse] Implement `nonisolated(nonsending)` type modifier

xedin committed:
[AST/ASTGen/Sema/Serialization] Remove `@execution` attribute

Complete the transition from `@execution` to `@concurrent` and `nonisolated(nonsending)`

xedin committed:
[Sema/SILGen] NFC: Remove all mentions of `@execution` from comments

xedin committed:
[AST] Remove `ExecutionAttribute` experimental feature

SE-0461 has been accepted and `@concurrent` and `nonisolated(nonsending)`
can be make generally available now.

xedin commented:
https://github.com/swiftlang/swift-syntax/pull/3047
@swift-ci please test

xedin commented:
https://github.com/swiftlang/swift-syntax/pull/3047
@swift-ci please test

hamishknight commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -542,8 +542,9 @@ ParserResult<Expr> Parser::parseExprSequenceElement(Diag<> message,
 
-  // Try to parse '@' sign or 'inout' as a attributed typerepr.
-  if (Tok.isAny(tok::at_sign, tok::kw_inout)) {
+  // Try to parse '@' sign, 'inout' or 'nonisolated' as a attributed typerepr.
+  if (Tok.isAny(tok::at_sign, tok::kw_inout) ||
+      Tok.isContextualKeyword("nonisolated")) {
```
Hmm, does this mean the following tries to parse as a `nonisolated` type?

```swift
func nonisolated(_ x: Int) {}
nonisolated(0)
print("hello")
```

I noticed the swift-syntax PR doesn't include similar logic, should this just be removed?

xedin replied:
If I remove this condition C++ parser cannot parse - `let _ = [nonisolated(nonsending) () async -> Void]()`, maybe there is a better way to do this I don't know about...

xedin replied:
I tightened the condition in `canParseNonisolatedAsTypeModifier` and added your snippet. Rintaro pointed me to this spot initially for `[nonisolated(nonsending) ...]()` support, but if there is a better way please let me know.

hamishknight reviewed (approved):
I have a question about the parsing logic in expression position, but otherwise this LGTM

xedin commented on include/swift/AST/Attr.h:
```diff
@@ -2978,29 +2974,38 @@ class ObjCImplementationAttr final : public DeclAttribute {
+    Bits.NonisolatedAttr.Modifier = static_cast<unsigned>(modifier);
   }
 
   NonisolatedAttr(bool unsafe, bool implicit)
```
I meant to do that in a follow-up but might as well :)

xedin replied:
✅

hamishknight replied:
Thanks!

xedin committed:
[AST] NFC: Add a convenient way to create implicit `NonisolatedAttr`s

xedin committed:
[Parse] Tighten constraints in `canParseNonisolatedAsTypeModifier`

Accept it only if it's spelled `nonisolated(nonsending)` and nothing
else to minimize any possible source compatibility impact.

xedin commented on include/swift/AST/Attr.h:
```diff
@@ -2978,29 +2974,38 @@ class ObjCImplementationAttr final : public DeclAttribute {
       : DeclAttribute(DeclAttrKind::Nonisolated, atLoc, range, implicit) {
-    Bits.NonisolatedAttr.isUnsafe = unsafe;
-    assert((isUnsafe() == unsafe) && "not enough bits for unsafe state");
+    Bits.NonisolatedAttr.Modifier = static_cast<unsigned>(modifier);
```
✅

xedin commented on lib/Parse/ParseType.cpp:
```diff
@@ -1730,13 +1734,51 @@ bool Parser::canParseTypeSimpleOrComposition() {
+    return false;
+
+  // Always requires `(<<argument>>)`
+  if (!Tok.is(tok::l_paren))
```
✅

hamishknight replied:
This doesn't seem to have changed?

xedin replied:
I replaced it with `consumeIf(tok::l_paren)`

xedin commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -2294,7 +2294,11 @@ ParamSpecifierRequest::evaluate(Evaluator &evaluator,
+
+  if (auto callerIsolated = dyn_cast<CallerIsolatedTypeRepr>(nestedRepr)) {
+    nestedRepr = callerIsolated->getBase();
+  }
```
✅

xedin commented:
https://github.com/swiftlang/swift-syntax/pull/3047
@swift-ci please test

AnthonyLatsis reviewed (approved):
Could you also update `userdocs/diagnostics/async-caller-execution.md`?

Curious: do we not need feature suppression for this new pair of attributes?

xedin commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -1414,10 +1414,8 @@ FunctionType::ExtInfo ClosureEffectsRequest::evaluate(
+  // `@concurrent` attribute is only valid on asynchronous function types.
+  if (expr->getAttrs().hasAttribute<ConcurrentAttr>()) {
     async = true;
   }
```
It's not applicable here.

xedin commented on include/swift/AST/Attr.h:
```diff
@@ -2978,14 +2978,9 @@ class NonisolatedAttr final : public DeclAttribute {
                   NonIsolatedModifier modifier, bool implicit)
       : DeclAttribute(DeclAttrKind::Nonisolated, atLoc, range, implicit) {
     Bits.NonisolatedAttr.Modifier = static_cast<unsigned>(modifier);
+    assert((getModifier() == modifier) && "not enough bits for modifier");
```
This was a suggestion by @hamishknight to end a round-trip here.

hamishknight replied:
IMO it's always nice to assert that a bitfield round-trips, yeah we count the bits, but this helps guard against bad refactorings or something else going awry with `SWIFT_INLINE_BITFIELD`, especially as the failure mode could otherwise be pretty subtle

xedin replied:
I'm just going to leave it as `assert` for now.

xedin commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -7550,6 +7556,20 @@ void AttributeChecker::visitNonisolatedAttr(NonisolatedAttr *attr) {
   // that do not have storage.
   auto dc = D->getDeclContext();
 
+  if (attr->isNonSending()) {
+    // Just like `@concurrent` this form of `nonisolated` is only
+    // applicable to certain declarations.
+    if (!(isa<FuncDecl>(D) || isa<ConstructorDecl>(D) ||
+          isa<SubscriptDecl>(D) || isa<VarDecl>(D))) {
+      diagnoseAndRemoveAttr(
+          attr, diag::cannot_specify_execution_behavior_for_decl, attr);
+      return;
+    }
```
This is not about `Concurrent` though and `nonisolated` is allowed in more more positions without arguments, what's why it's this way...

xedin replied:
Actually maybe that's a good way to spell that since we are going to keep them consistent...

xedin replied:
It's that way today but if we ever make `@concurrent` behave different it would be a problem, I made a change but we'd need to keep an eye...

xedin commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -2172,5 +2172,18 @@ ERROR(sil_thunkinst_failed_to_parse_kind,none,
+ERROR(nonisolated_nonsending_incorrect_modifier,PointsToFirstBadToken,
+      "expected 'nonsending' in modifier", ())
+ERROR(nonisolated_nonsending_expected_rparen,PointsToFirstBadToken,
+      "expected '(' after 'nonisolated' modifier", ())
```
Oops, thanks!

xedin replied:
✅

xedin commented on lib/SIL/IR/SILFunctionType.cpp:
```diff
@@ -2622,15 +2622,17 @@ static CanSILFunctionType getSILFunctionType(
-        switch (*decl->getExecutionBehavior()) {
-        case ExecutionKind::Concurrent:
+      } else if (auto *decl = constant->getAbstractFunctionDecl()) {
+        if (auto behavior = decl->getExecutionBehavior()) {
```
Yes, the flow is different now, we need to check different attributes.

AnthonyLatsis replied:
What I found concerning is that currently a non-null `AbstractFunctionDecl` without `@execution` will cause control to enter the next `else` branch where isolation is set to `getActorIsolationOfContext`, but your change leaves the isolation nil for `AbstractFunctionDecl` whenever it is not determined by either of the new attributes.

xedin replied:
Oh, I see what you mean, I'll fix that.

xedin replied:
✅

xedin commented:
> Curious: do we not need feature suppression for this new pair of attributes?

We don't need a feature for this since the proposal was been accepted, that's why I've removed `ExecutionAttribute` feature in this PR.

xedin committed:
[Diagnostics] Update educational note for `AsyncCallerExecution` feature to match `@execution` attribute split

AnthonyLatsis commented:
> We don't need a feature for this since the proposal was been accepted

I thought the practice was to ship syntactic features as suppressible for a while for short-term textual interface compatibility.

xedin commented:
These new attribute and modifier are not going to be present in old interface files unconditionally anyway.

tshortli commented:
No, Anthony is right. There is a trailing window of previous development versions of the compiler that textual interfaces must remain compatible with.

xedin committed:
[Frontend] Add a feature to guard use of `@concurrent` and `nonisolated(nonsending)` in swift interface files

xedin commented:
For posterity: we discussed this offline and yes, we do need a feature flag here to indicate that new attributes are supported only by new compilers even though we have been inconsistent about this in the past.

xedin commented:
https://github.com/swiftlang/swift-syntax/pull/3047
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
https://github.com/swiftlang/swift-syntax/pull/3047
@swift-ci please test source compatibility

xedin commented:
https://github.com/swiftlang/swift-syntax/pull/3047
@swift-ci please test macOS platform
